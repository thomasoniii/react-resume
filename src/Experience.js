import React from 'react';

import {PROJECT_BLURBS} from './filter_types';
import './styles/Experience.css';

const renderHighlight = (highlight, filters) => {
console.log(filters, highlight.tech, highlight.blurb);
  if (Object.keys(filters).length && highlight.tech) {
    console.log("DISPLAY TECH : ", highlight.tech, filters);
    let shouldDisplay = highlight.tech.reduce( (sd, f) => {return sd || filters[f]}, false);
    if (!shouldDisplay) { return null };
  }

  return (
    <li key={highlight.blurb}>
      { highlight.blurb }
      { renderTech(highlight.tech, filters) }
    </li>
  )
}

const renderTech = (tech = [], filters) => {
  if (! tech.length) { return null }

  return (
    <div className='project-tech-box'>
      <span className='subsubheading'>Technology used:</span>
      <span className='project-tech'>{tech.join(', ')}</span>
    </div>
  )
}

export default ({experience, filters : all_filters, sections, projects}) => {
  let filters = Object.keys(all_filters).reduce( (acc, f) => { if (all_filters[f]) { acc[f] = true }; return acc }, {});
  console.log("BLURBS : ", filters, PROJECT_BLURBS, filters[PROJECT_BLURBS]);
  return (
    <div className='container-fluid section experience'>
      <div className='row'>
        <div className='col section-header'>
          Experience
        </div>
      </div>
      { experience.map( job => {
        return [
          <div className='row job-company' key='name'>
            <div className='col-sm'>
              <span className='employer'>{job.employer}</span>
              <span className='location'>, {job.location}</span>
            </div>
            <div className='col-sm duration'>
              <div className='float-right'>{job.duration.from} - {job.duration.to}</div>
            </div>
          </div>,
          job.title && job.title.map( role => {
            return (
              <div className='row' key={role}>
                <div className='col job-role'>
                  { role }
                </div>
              </div>
            )
          }),
          job.projects && !job.collapsed && job.projects.map( project => {
            const project_name = project.project;
            return [
              <div className='row job-project' key={`${project_name}-info`}>
                <div className='col-md'>
                  <span className='project-name'>{project_name}</span>
                  <span className='project-role'>{project.role}</span>
                  <span className='project-members'>{project.members} person team</span>
                  {project.url && (
                    <span className='float-right'>
                      { project.url.map(url => <a href = {url} target='_blank'> {url}</a>) }
                    </span>
                  )}
                </div>
                <div className='col-md date order-sm-last'>
                  <div className='float-right'>
                    {project.date}
                  </div>
                </div>
              </div>,
              projects[project_name] && sections[PROJECT_BLURBS] && <div className='row' key={`${project_name}-blurb`}>
                <div className='col-md project-blurb order-md-first'>
                  { projects[project_name].blurb }
                </div>
                {projects[project_name].url && (
                  <div className='col-md order-first'>
                    <ul style={{listStyle : 'none'}} className='float-right'>
                      {projects[project_name].url.map(url => {
                        return <li key={url}><a href = {url} target='_blank'>{url}</a></li>
                      })}
                    </ul>
                  </div>
                )}
              </div>,
              project.highlights && !project.collapsed && (
                <div className='row project-highlight' key={`${project_name}-highlights`}>
                  <div className='col highlight'>
                    <ul>
                      {project.highlights.map(highlight => {
                        return renderHighlight(highlight, filters)
                      })}
                    </ul>
                  </div>
                </div>
              ),
            ]
          })
        ]
      })}
    </div>
  )
}
