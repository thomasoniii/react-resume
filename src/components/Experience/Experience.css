.job-container,
.job-project-container {
  display: flex;
  justify-content: space-between;
  cursor: pointer;
}

.job-container:hover,
.job-project-container:hover {
  background-color: lightgray;
}

.highlight ul {
  padding-left: 15px;
}

.results {
  font-style: italic;
}

.tech {
  font-style: italic;
  color: gray;
}

.duration {
  font-style: italic;
}

.employer {
  font-weight: bold;
  text-decoration: underline;
}

.job-project {
  margin-left: 2vh;
}

.job-role {
  margin-left: 1vh;
  font-weight: bold;
}

.project-name {
  font-weight: bold;
}

.job-project-name {
  font-style: italic;
}

.job-project-name::after {
  content: " - ";
}

.project-role {
  text-decoration: underline;
}

.project-members::before {
  content: ", ";
}

.project-highlight {
  margin-left: 3vh;
}

.project-tech {
  font-style: italic;
  color: gray;
}

.project-blurb {
  font-style: italic;
  font-size: 95%;
  margin-left: 4vh;
}

.subsubheading {
  text-decoration: underline;
}

.job-company,
.job-project {
  cursor: pointer;
}

.job-company:hover,
.job-project:hover {
  background-color: lightgray;
}
