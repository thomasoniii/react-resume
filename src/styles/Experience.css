.duration {
  font-style : italic;
}

.employer {
  font-weight : bold;
  text-decoration : underline;
}

.job-project {
  margin-left : 2vh;
}

.job-role {
  margin-left : 1vh;
  font-weight : bold;
}

.project-name {
  font-weight : bold;
}

.project-name::after {
  content : ' - '
}

.project-role {
  text-decoration : underline;
}

.project-members::before {
  content : ', '
}

.project-highlight {
  margin-left : 3vh;
}

.project-tech {
  font-style : italic;
  color : gray;
}

.project-blurb {
  font-style : italic;
  font-size : 95%;
  margin-left : 4vh;
}

.subsubheading {
  text-decoration : underline;
}

.job-company:hover {
  background-color : lightgray;
}

.job-project:hover {
  background-color : lightgray;
}
