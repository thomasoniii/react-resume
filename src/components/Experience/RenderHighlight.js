import React from "react";

import Tech from "components/Tech";

import "./Experience.css";

const NonMemoRenderTech = ({ tech = [] }) => {
  if (!tech.length) {
    return null;
  }

  return (
    <div className="project-tech-box">
      <Tech tech={tech} />
    </div>
  );
};

const RenderTech = React.memo(NonMemoRenderTech);

const NonMemoRenderHighlight = ({ highlight, filters = [] }) => {
  if (
    filters.length &&
    highlight.tech &&
    !highlight.tech.some((t) => filters.includes(t))
  ) {
    return null;
  }

  return (
    <li key={highlight.blurb}>
      {highlight.blurb}
      <RenderTech tech={highlight.tech} />
    </li>
  );
};

const RenderHighlight = React.memo(NonMemoRenderHighlight);

export default RenderHighlight;
