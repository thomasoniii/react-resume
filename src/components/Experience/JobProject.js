import React from "react"
import PropTypes from "prop-types"

import { PROJECT_BLURBS } from "filter_types"
import RenderHighlight from "./RenderHighlight"

const JobProject = ({
  project,
  collapseCallback,
  collapsed,
  sections,
  projects,
  filters,
}) => {
  const project_name = project.project
  return (
    <React.Fragment key={project_name}>
      <div
        className="job-project"
        key={`${project_name}-info`}
        onClick={() => {
          collapseCallback(project_name)
        }}
      >
        <div className="job-project-container">
          <span>
            <span className="job-project-name">{project_name}</span>
            <span className="project-role">{project.role}</span>

            <span className="project-members">
              {project.members} person team
            </span>
          </span>
          {project.url && (
            <span className="float-right">
              {project.url.map((url) => (
                <a
                  href={url}
                  target="_blank"
                  rel="noopener noreferrer"
                  key={url}
                >
                  {" "}
                  {url}
                </a>
              ))}
            </span>
          )}

          <div className="date">
            <div className="float-right">{project.date}</div>
          </div>
        </div>
      </div>
      {projects[project_name] && sections[PROJECT_BLURBS] && (
        <div key={`${project_name}-blurb`}>
          <div className="project-blurb">{projects[project_name].blurb}</div>
          {projects[project_name].url && (
            <div>
              <ul style={{ listStyle: "none" }} className="float-right">
                {projects[project_name].url.map((url) => {
                  return (
                    <li key={url}>
                      <a href={url} target="_blank" rel="noopener noreferrer">
                        {url}
                      </a>
                    </li>
                  )
                })}
              </ul>
            </div>
          )}
        </div>
      )}
      {project.highlights && !collapsed[project.project] && (
        <div className="project-highlight" key={`${project_name}-highlights`}>
          <div className="highlight">
            <ul>
              {project.highlights.map((highlight, i) => {
                return (
                  <RenderHighlight
                    key={i}
                    highlight={highlight}
                    filters={filters}
                  />
                )
              })}
            </ul>
          </div>
        </div>
      )}
    </React.Fragment>
  )
}

JobProject.propTypes = {
  project: PropTypes.object,
  collapseCallback: PropTypes.func,
  collapsed: PropTypes.bool,
  sections: PropTypes.array,
  projects: PropTypes.object,
  filters: PropTypes.array,
}

export default React.memo(JobProject)
