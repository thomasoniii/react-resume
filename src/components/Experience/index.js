import React from "react"
import PropTypes from "prop-types"
import { Card, Typography, Timeline } from "antd"

import JobProject from "./JobProject"

import "./Experience.css"

const { Title } = Typography

const Experience = ({
  experience,
  tech_filters,
  sections,
  projects,
  collapsed,
  collapseCallback,
}) => {
  return (
    <Card size="small" title={<Title level={2}>Experience</Title>}>
      <Timeline>
        {experience.map((job) => (
          <Timeline.Item
            key={job.id}
            color={job.duration.to === "present" ? "green" : undefined}
          >
            <div
              className="job-container"
              onClick={() => {
                collapseCallback(job.id)
              }}
            >
              <span>
                <span className="job-name">
                  <span className="employer">{job.employer}</span>
                  <span className="location">, {job.location}</span>
                </span>
                {job.url && (
                  <a
                    href={job.url}
                    onClick={(e) => e.stopPropagation()}
                    target="_blank"
                    rel="noopener noreferrer"
                    className="job-link"
                  >
                    {job.url}
                  </a>
                )}
              </span>
              <span className="duration align-right">
                {job.duration.from} - {job.duration.to}
              </span>
            </div>
            {job.title &&
              job.title.map((role) => (
                <div key={role}>
                  <div className="job-role">{role}</div>
                </div>
              ))}
            {job.projects &&
              !collapsed[job.id] &&
              job.projects.map((project) => (
                <JobProject
                  key={project.project}
                  project={project}
                  collapsed={collapsed}
                  sections={sections}
                  projects={projects}
                  collapseCallback={collapseCallback}
                  filters={tech_filters}
                />
              ))}
          </Timeline.Item>
        ))}
      </Timeline>
    </Card>
  )
}

Experience.propTypes = {
  experience: PropTypes.array,
  tech_filters: PropTypes.array,
  sections: PropTypes.array,
  projects: PropTypes.object,
  collapsed: PropTypes.object,
  collapseCallback: PropTypes.func,
}

export default React.memo(Experience)
