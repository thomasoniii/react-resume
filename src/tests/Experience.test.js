import React from "react"
import ReactDOM from "react-dom"
import renderer from "react-test-renderer"

import Experience from "../components/Experience"
import resume from "../data/resume-example.json"

/*
<Experience
  filters={ filters }
  experience={resume.experience}
  projects={resume.projects}
  sections={sections}
  collapsed={collapsed}
  collapseCallback={this.props.toggleCollapsed}
  key='experience'/>
*/

const filters = {}
const collapsed = {}
const collapsedCallback = () => {}
const sections = {}

describe("Experience tests", () => {
  test("Experience exists properly", () => {
    const experience = renderer.create(
      <Experience
        filters={filters}
        experience={resume.experience}
        projects={resume.projects}
        sections={sections}
        collapsed={collapsed}
        collapseCallback={collapsedCallback}
      />
    )

    const tree = experience.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
