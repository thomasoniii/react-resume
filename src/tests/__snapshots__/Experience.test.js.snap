// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Experience tests Experience exists properly 1`] = `
<div
  className="container-fluid section experience"
>
  <div
    className="row"
  >
    <div
      className="col section-header"
    >
      Experience
    </div>
  </div>
  <div
    className="row job-company"
    onClick={[Function]}
  >
    <div
      className="col-sm"
    >
      <span
        className="employer"
      >
        Awesome Co
      </span>
      <span
        className="location"
      >
        , 
        Awesome Town, USA
      </span>
    </div>
    <div
      className="col-sm duration"
    >
      <div
        className="float-right"
      >
        January 2017
         - 
        present
      </div>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col job-role"
    >
      Awesome Employee
    </div>
  </div>
  <div
    className="row job-project"
    onClick={[Function]}
  >
    <div
      className="col-md"
    >
      <span
        className="job-project-name"
      >
        Awesome Sauce
      </span>
      <span
        className="project-role"
      >
        Sauce Chef
      </span>
      <span
        className="project-members"
      >
        5
         person team
      </span>
    </div>
    <div
      className="col-md date order-sm-last"
    >
      <div
        className="float-right"
      >
        2018-present
      </div>
    </div>
  </div>
  <div
    className="row project-highlight"
  >
    <div
      className="col highlight"
    >
      <ul>
        <li>
          Here is where you can describe (in multiple blurbs, with multiple techs), the awesome things you did to contribute to this project.
          <div
            className="project-tech-box"
          >
            <span
              className="subsubheading"
            >
              Technology used:
            </span>
            <span
              className="project-tech"
            >
              Tech I know
            </span>
          </div>
        </li>
        <li>
          This is another awesome thing I did while making Awesome Sauce. It used different tech.
          <div
            className="project-tech-box"
          >
            <span
              className="subsubheading"
            >
              Technology used:
            </span>
            <span
              className="project-tech"
            >
              Other Tech I know
            </span>
          </div>
        </li>
        <li>
          This is another thing I did that won't be filtered by tech. You can click on anything that highlights in gray on hover to collapse it.
        </li>
      </ul>
    </div>
  </div>
  <div
    className="row job-project"
    onClick={[Function]}
  >
    <div
      className="col-md"
    >
      <span
        className="job-project-name"
      >
        Awesome Marmalade
      </span>
      <span
        className="project-role"
      >
        Marmalade Wrangler
      </span>
      <span
        className="project-members"
      >
        8
         person team
      </span>
    </div>
    <div
      className="col-md date order-sm-last"
    >
      <div
        className="float-right"
      >
        2017-2018
      </div>
    </div>
  </div>
  <div
    className="row project-highlight"
  >
    <div
      className="col highlight"
    >
      <ul>
        <li>
          This is the first project I did at AwesomeCo. I was so awesome I got moved to making Awesome Sauce.
          <div
            className="project-tech-box"
          >
            <span
              className="subsubheading"
            >
              Technology used:
            </span>
            <span
              className="project-tech"
            >
              Tech I know
            </span>
          </div>
        </li>
        <li>
          My next job was far enough in the past, doing different things, that I collapsed it by default. Click on it to expand.
          <div
            className="project-tech-box"
          >
            <span
              className="subsubheading"
            >
              Technology used:
            </span>
            <span
              className="project-tech"
            >
              Tech I know
            </span>
          </div>
        </li>
      </ul>
    </div>
  </div>
  <div
    className="row job-company"
    onClick={[Function]}
  >
    <div
      className="col-sm"
    >
      <span
        className="employer"
      >
        That Other Company
      </span>
      <span
        className="location"
      >
        , 
        Other Town, USA
      </span>
    </div>
    <div
      className="col-sm duration"
    >
      <div
        className="float-right"
      >
        May 2015
         - 
        December 2016
      </div>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col job-role"
    >
      Wage Drone
    </div>
  </div>
  <div
    className="row job-project"
    onClick={[Function]}
  >
    <div
      className="col-md"
    >
      <span
        className="job-project-name"
      >
        Nothing Special
      </span>
      <span
        className="project-role"
      >
        Wage Monkey
      </span>
      <span
        className="project-members"
      >
        50
         person team
      </span>
    </div>
    <div
      className="col-md date order-sm-last"
    >
      <div
        className="float-right"
      >
        2015-2016
      </div>
    </div>
  </div>
  <div
    className="row project-highlight"
  >
    <div
      className="col highlight"
    >
      <ul>
        <li>
          I was a cog in the machine, so I kept this collapsed by default. This just shows that I had a job, but it doesn't matter currently. The next project was really unimportant, so it's also collapsed.
        </li>
      </ul>
    </div>
  </div>
  <div
    className="row job-project"
    onClick={[Function]}
  >
    <div
      className="col-md"
    >
      <span
        className="job-project-name"
      >
        Kitchen cleaner
      </span>
      <span
        className="project-role"
      >
        Scullery maid
      </span>
      <span
        className="project-members"
      >
        1
         person team
      </span>
    </div>
    <div
      className="col-md date order-sm-last"
    >
      <div
        className="float-right"
      >
        May 2015-June 2015
      </div>
    </div>
  </div>
  <div
    className="row project-highlight"
  >
    <div
      className="col highlight"
    >
      <ul>
        <li>
          It's embarrassing, but I helped clean up our fridge. Which was important. I kept this here to show that I'll do unpleasant tasks for the company, but didn't want to highlight it.
        </li>
      </ul>
    </div>
  </div>
</div>
`;
